### PR TITLE
Reader: defend against invalid URL keys

### PR DESCRIPTION
**Pull request by bluefuton**
While working with subscriptions in state, I noticed I had deleted sites in my `state.reader.follows.items` with the keys `www` and `www.`. 

This PR adds checks for these common invalid keys, and as a result drops any subscriptions that have an empty feed URL.

### To test

Clear your local storage. 

Try subscribing and unsubscribing to feeds, and loading your subscriptions at http://calypso.localhost:3000/following/manage. Check there are no regressions.

Type `state.reader.follows.items` into your console and check that there are no subscriptions keyed off 'www' or 'www.'.

### CODE REVIEW

**Comment by samouri** [comment]
Why are we dispatching an empty feedUrl to begin with?

**Comment by bluefuton** [comment]
> Do you know why are we dispatching an empty feedUrl to begin with?

I think it happens when we fetch the list of subscriptions and are handed some dodgy data from the endpoint (including deleted sites). Ideally we'll fix up the endpoint, but this protects us against it happening for the moment.

File: client/state/reader/follows/utils.js [hunk 1]
```diff
@@ -5,6 +5,14 @@
...
 
 export function prepareComparableUrl( url ) {
-	const preparedUrl = url && untrailingslashit( url );
-	return preparedUrl && preparedUrl.replace( /^https?:\/\//, '' ).toLowerCase();
+	const unslashedUrl = url && untrailingslashit( url );
+	if ( ! unslashedUrl ) {
+		return null;
+	}
+	const preparedUrl = unslashedUrl.replace( /^https?:\/\//, '' ).toLowerCase();
+	if ( preparedUrl === 'www' || preparedUrl === 'www.' ) {
```
Comment (blowery) [inline client/state/reader/follows/utils.js]: maybe we should expand this to catch more invalid things? We could only let strings that contain `\w+\.\w+` (anything followed by a `.` followed by anything) through...

That would also catch things like `localhost` and weird singular names that only work on intranets

Comment (bluefuton) [reply client/state/reader/follows/utils.js]: Just remembered that we have `resemblesUrl` in `lib/url` for exactly this case. I'll pop that in.

**Comment by Copons** [comment]
Code LGTM here, but I'm not sure how to make it fail on current master. 🤔 
I couldn't find any feed with the URL `www` (although I've got plenty _starting_ with that).

**Comment by bluefuton** [comment]
Here are a couple of feeds without valid URLs:

https://wpcalypso.wordpress.com/read/feeds/10432419
https://wpcalypso.wordpress.com/read/feeds/9236936

**Comment by codebykat** [comment]
I confirmed that following one of the above feeds on master added a stray `www` key in `state.reader.follows.items`, and that the key is not removed when you unfollow the site.

On this branch, I can't follow the feed at all. Is that the intended behavior? It feels broken to click the Follow button and have no feedback whatsoever. (It does succeed in following behind the scenes, as when I switch back to master, it's in the list again.)

If I had previously followed one of the broken feeds, switching to this branch removes it from my Following list and `state.reader.follows.items`.

**Comment by Copons** [comment]
@codebykat For what I understand, that's the expected behaviour.
Like, these are leftovers of deleted sites, not something we would want our users to follow in the Reader. 🙂

**Comment by bluefuton** [comment]
On reflection, I think Calypso is probably behaving the right way here. It's trusting what it gets back from the API, and these feeds do exist in Feedbag. As @codebykat mentions, if we exclude them from state altogether, we get other undesirable behaviour (the follow button fails with no feedback).

I think we should focus on fixing this on the backend.

**Pull request closed without merging**